src/tests.c

```diff
@@ -6712,34 +6590,30 @@ static void run_pubkey_comparison(void) {
     };
     secp256k1_pubkey pk1;
     secp256k1_pubkey pk2;
-    int32_t ecount = 0;
 
     CHECK(secp256k1_ec_pubkey_parse(CTX, &pk1, pk1_ser, sizeof(pk1_ser)) == 1);
     CHECK(secp256k1_ec_pubkey_parse(CTX, &pk2, pk2_ser, sizeof(pk2_ser)) == 1);
 
-    secp256k1_context_set_illegal_callback(CTX, counting_illegal_callback_fn, &ecount);
-    CHECK(secp256k1_ec_pubkey_cmp(CTX, NULL, &pk2) < 0);
-    CHECK(ecount == 1);
-    CHECK(secp256k1_ec_pubkey_cmp(CTX, &pk1, NULL) > 0);
-    CHECK(ecount == 2);
+    CHECK_ILLEGAL_VOID(CTX, CHECK(secp256k1_ec_pubkey_cmp(CTX, NULL, &pk2) < 0));
+    CHECK_ILLEGAL_VOID(CTX, CHECK(secp256k1_ec_pubkey_cmp(CTX, &pk1, NULL) > 0));
     CHECK(secp256k1_ec_pubkey_cmp(CTX, &pk1, &pk2) < 0);
     CHECK(secp256k1_ec_pubkey_cmp(CTX, &pk2, &pk1) > 0);
     CHECK(secp256k1_ec_pubkey_cmp(CTX, &pk1, &pk1) == 0);
     CHECK(secp256k1_ec_pubkey_cmp(CTX, &pk2, &pk2) == 0);
-    CHECK(ecount == 2);
     {
         secp256k1_pubkey pk_tmp;
         memset(&pk_tmp, 0, sizeof(pk_tmp)); /* illegal pubkey */
-        CHECK(secp256k1_ec_pubkey_cmp(CTX, &pk_tmp, &pk2) < 0);
-        CHECK(ecount == 3);
-        CHECK(secp256k1_ec_pubkey_cmp(CTX, &pk_tmp, &pk_tmp) == 0);
-        CHECK(ecount == 5);
-        CHECK(secp256k1_ec_pubkey_cmp(CTX, &pk2, &pk_tmp) > 0);
-        CHECK(ecount == 6);
+        CHECK_ILLEGAL_VOID(CTX, CHECK(secp256k1_ec_pubkey_cmp(CTX, &pk_tmp, &pk2) < 0));
+        {
+            int32_t ecount = 0;
+            secp256k1_context_set_illegal_callback(CTX, counting_illegal_callback_fn, &ecount);
+            CHECK(secp256k1_ec_pubkey_cmp(CTX, &pk_tmp, &pk_tmp) == 0);
+            CHECK(ecount == 2);
+            secp256k1_context_set_illegal_callback(CTX, NULL, NULL);
```

Comment:
Is it a bug that `secp256k1_ec_pubkey_cmp` calls the callback twice?

I mean, we never do this anywhere else, even if multiple arguments are illegal. In particular, `ARG_CHECK` makes the function return immediately on failure.

Reply:
It's intentional, see the comment in `pubkey_cmp`:

```
        /* If the public key is NULL or invalid, ec_pubkey_serialize will call
         * the illegal_callback and return 0. In that case we will serialize the
         * key as all zeros which is less than any valid public key. This
         * results in consistent comparisons even if NULL or invalid pubkeys are
         * involved and prevents edge cases such as sorting algorithms that use
         * this function and do not terminate as a result. */
```

In particular, returning 0 immediately in pubkey_cmp if parsing fails does not signal an error to the color, but that both public keys are identical.

Reply:
Oh, crazy. I still think we could get the best of both worlds and retain the current all-zeros behavior and at the same time call the callback only once. But that's better done in a separate PR then.